tsfc/kernel_args.py

```diff
@@ -0,0 +1,76 @@
+import abc
+
+import coffee.base as coffee
+import loopy as lp
+
+
+class KernelArg(abc.ABC):
+    """Abstract base class wrapping either a COFFEE or loopy argument.
+
+    Defining this type system allows Firedrake (or other libraries) to
+    prepare arguments for the kernel without needing to worry about their
+    ordering. Instead this can be offloaded to tools such as
+    :func:`functools.singledispatch`.
+    """
+
+    def __init__(self, ast_arg):
+        self._ast_arg = ast_arg
+
+    @property
+    def dtype(self):
+        if self._is_coffee_backend:
+            return self._ast_arg.typ
+        elif self._is_loopy_backend:
+            return self._ast_arg.dtype
+
+    @property
+    def coffee_arg(self):
+        if not self._is_coffee_backend:
+            raise RuntimeError("Invalid type requested")
+        return self._ast_arg
+
+    @property
+    def loopy_arg(self):
+        if not self._is_loopy_backend:
+            raise RuntimeError("Invalid type requested")
+        return self._ast_arg
+
+    @property
+    def _is_coffee_backend(self):
+        return isinstance(self._ast_arg, coffee.Decl)
+
+    @property
+    def _is_loopy_backend(self):
+        return isinstance(self._ast_arg, lp.ArrayArg)
+
+
+class OutputKernelArg(KernelArg):
+    ...
+
+
+class CoordinatesKernelArg(KernelArg):
+    ...
+
+
+class CoefficientKernelArg(KernelArg):
+    ...
+
+
+class CellOrientationsKernelArg(KernelArg):
+    ...
+
+
+class CellSizesKernelArg(KernelArg):
+    ...
+
+
+class TabulationKernelArg(KernelArg):
+    ...
+
+
+class ExteriorFacetKernelArg(KernelArg):
+    ...
+
+
+class InteriorFacetKernelArg(KernelArg):
+    ...
```

Comment:
Would it make sense for these to all be enum instances, rather than classes? Then you can handle the cases in a match/case and static analysis can say something useful about whether everything is handled.

Reply:
Possibly. I chose this route for two reasons:
- I can use `functools.singledispatch` to handle them.
- I can imagine a future where TSFC might attach more functionality to these classes (e.g. to generate the loopy arg from a FInAT element).